src/python-bindings/classad2/exprtree.cpp

```diff
@@ -1 +1,155 @@
+static PyObject *
+_exprtree_init(PyObject *, PyObject * args) {
+	// _exprtree_init( self, self._handle )
 
+    PyObject * self = NULL;
+    PyObject_Handle * handle = NULL;
+    if(! PyArg_ParseTuple( args, "OO", & self, (PyObject **)& handle )) {
+        // PyArg_ParseTuple() has already set an exception for us.
+        return NULL;
+    }
+
+    // We don't know which kind of ExprTree this is going to be yet,
+    // so we can't initialize it to anything (ExprTree proper is abstract).
+    handle->t = NULL;
+    handle->f = [](void *& v) { dprintf( D_PERF_TRACE, "[unconstructed ExprTree]\n" ); if( v != NULL ) { dprintf(D_ALWAYS, "Error!  Unconstructed ExprTree has non-NULL handle %p\n", v); } };
+    Py_RETURN_NONE;
+}
+
+
+static PyObject *
+_exprtree_eq( PyObject *, PyObject * args ) {
+    // _exprtree_eq( self._handle, other._handle )
+
+    PyObject_Handle * self = NULL;
+    PyObject_Handle * other = NULL;
+    if(! PyArg_ParseTuple( args, "OO", (PyObject **)& self, (PyObject **)& other)) {
+        // PyArg_ParseTuple() has already set an exception for us.
+        return NULL;
+    }
+
+    if( *(classad::ExprTree *)self->t == *(classad::ExprTree *)other->t ) {
+        Py_RETURN_TRUE;
+    } else {
+        Py_RETURN_FALSE;
+    }
+}
+
+
+bool
+EvaluateLooseExpr( classad::ExprTree * expr,
+                   classad::ClassAd * my, classad::ClassAd * target,
+                   classad::Value & value ) {
+    const classad::ClassAd * originalScope = expr->GetParentScope();
+    expr->SetParentScope(my);
+
+    bool rv = false;
+    if( target == my || target == NULL ) {
+        rv = expr->Evaluate( value );
+    } else {
+        classad::MatchClassAd matchAd( my, target );
+        rv = expr->Evaluate( value );
+        matchAd.RemoveLeftAd();
+        matchAd.RemoveRightAd();
+    }
+
+    expr->SetParentScope(originalScope);
+    return rv;
+}
+
+
+bool
+evaluate( classad::ExprTree * expr, classad::ClassAd * scope,
+  classad::ClassAd * target, classad::Value & value
+) {
+    if( scope != NULL ) {
+        return EvaluateLooseExpr( expr, scope, target, value );
+    } else if( expr->GetParentScope() ) {
+        return expr->Evaluate( value );
+    } else {
+        classad::EvalState state;
+        return expr->Evaluate( state, value );
+    }
```

Comment:
These two blocks are equivalent.

Reply:
The ClassAd API claims that they aren't.  I don't know what happens if you call `state.SetScopes( GetParentScope() )` when there's no parent scope -- it looks safe based a on quick code inspection -- but I'd rather not worry about it.